diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2011-10-28 14:34:49.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2011-10-28 14:39:27.000000000 +0530
@@ -118,6 +118,7 @@ const struct mcu_type_s avr_mcu_types[] 
     /* Enhanced, <= 8K + 2-byte PC + { MOVW/LPMX, MUL }.  */
   { "avr4",                 ARCH_AVR4, NULL,                        0, 0x0060, "m8" },
   { "atmega8",              ARCH_AVR4, "__AVR_ATmega8__",           0, 0x0060, "m8" },
+  { "atmega8a",             ARCH_AVR4, "__AVR_ATmega8A__",          0, 0x0060, "m8a" },
   { "atmega48",             ARCH_AVR4, "__AVR_ATmega48__",          0, 0x0100, "m48" },
   { "atmega48a",            ARCH_AVR4, "__AVR_ATmega48A__",         0, 0x0100, "m48a" },
   { "atmega48pa",           ARCH_AVR4, "__AVR_ATmega48PA__",        0, 0x0100, "m48pa" },
@@ -242,8 +243,10 @@ const struct mcu_type_s avr_mcu_types[] 
        { MOVW/LPMX, JMP/CALL, MUL, ELPM, ELPMX }.  */
   { "avr51",                ARCH_AVR51, NULL,                       0, 0x0100, "m128" },
   { "atmega128",            ARCH_AVR51, "__AVR_ATmega128__",        0, 0x0100, "m128" },
+  { "atmega128a",           ARCH_AVR51, "__AVR_ATmega128A__",       0, 0x0100, "m128a" },
   { "atmega1280",           ARCH_AVR51, "__AVR_ATmega1280__",       0, 0x0200, "m1280" },
   { "atmega1281",           ARCH_AVR51, "__AVR_ATmega1281__",       0, 0x0200, "m1281" },
+  { "atmega1284",           ARCH_AVR51, "__AVR_ATmega1284__",       0, 0x0100, "m1284" },
   { "atmega1284p",          ARCH_AVR51, "__AVR_ATmega1284P__",      0, 0x0100, "m1284p" },
   { "atmega128rfa1",        ARCH_AVR51, "__AVR_ATmega128RFA1__",    0, 0x0200, "m128rfa1" },
   { "atmega128rfa2",        ARCH_AVR51, "__AVR_ATmega128RFA2__",    0, 0x0200, "m128rfa2" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2011-10-28 14:55:08.000000000 +0530
+++ gcc/config/avr/t-avr	2011-10-28 14:52:10.000000000 +0530
@@ -157,6 +157,7 @@ MULTILIB_MATCHES = \
 	mmcu?avr4=mmcu?atmega48pa \
 	mmcu?avr4=mmcu?atmega48p \
 	mmcu?avr4=mmcu?atmega8 \
+	mmcu?avr4=mmcu?atmega8a \
 	mmcu?avr4=mmcu?atmega8515 \
 	mmcu?avr4=mmcu?atmega8535 \
 	mmcu?avr4=mmcu?atmega88 \
@@ -263,8 +264,10 @@ MULTILIB_MATCHES = \
 	mmcu?avr5=mmcu?at94k \
 	mmcu?avr5=mmcu?m3000 \
 	mmcu?avr51=mmcu?atmega128 \
+	mmcu?avr51=mmcu?atmega128a \
 	mmcu?avr51=mmcu?atmega1280 \
 	mmcu?avr51=mmcu?atmega1281 \
+	mmcu?avr51=mmcu?atmega1284 \
 	mmcu?avr51=mmcu?atmega1284p \
 	mmcu?avr51=mmcu?atmega128rfa1 \
 	mmcu?avr51=mmcu?atmega128rfa2 \
