diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2011-10-28 15:04:23.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2011-10-28 15:06:36.000000000 +0530
@@ -265,6 +265,7 @@ const struct mcu_type_s avr_mcu_types[] 
     /* Xmega, <= 8K FLASH.  */
     /* Xmega, > 8K, <= 64K FLASH, <= 64K RAM.  */
   { "avrxmega2",    ARCH_AVRXMEGA2, NULL,                           0, 0x2000, "x32a4" },
+  { "atmxt336s",    ARCH_AVRXMEGA2, "__AVR_ATMXT336S__",            0, 0x2000, "mxt336s"},
   { "atxmega16a4",  ARCH_AVRXMEGA2, "__AVR_ATxmega16A4__",          0, 0x2000, "x16a4" },
   { "atxmega16a4u",  ARCH_AVRXMEGA2, "__AVR_ATxmega16A4U__",        0, 0x2000, "x16a4u" },
   { "atxmega16d4",  ARCH_AVRXMEGA2, "__AVR_ATxmega16D4__",          0, 0x2000, "x16d4" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2011-10-28 15:04:23.000000000 +0530
+++ gcc/config/avr/t-avr	2011-10-28 15:06:36.000000000 +0530
@@ -279,6 +279,7 @@ MULTILIB_MATCHES = \
 	mmcu?avr6=mmcu?atmega256rfr2 \
  	mmcu?avr6=mmcu?atmega2560 \
 	mmcu?avr6=mmcu?atmega2561 \
+	mmcu?avrxmega2=mmcu?atmxt336s \
 	mmcu?avrxmega2=mmcu?atxmega16a4 \
 	mmcu?avrxmega2=mmcu?atxmega16a4u \
 	mmcu?avrxmega2=mmcu?atxmega16d4 \
