diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2011-11-08 15:04:32.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2011-11-08 15:06:12.000000000 +0530
@@ -310,6 +310,7 @@ const struct mcu_type_s avr_mcu_types[] 
   { "atxmega256c3", ARCH_AVRXMEGA6, "__AVR_ATxmega256C3__",         0, 0x2000, "x256c3" },
   { "atxmega256d3", ARCH_AVRXMEGA6, "__AVR_ATxmega256D3__",         0, 0x2000, "x256d3" },
   { "atxmega384c3", ARCH_AVRXMEGA6, "__AVR_ATxmega384C3__",         0, 0x2000, "x384c3" },
+  { "atxmega384d3", ARCH_AVRXMEGA6, "__AVR_ATxmega384D3__",         0, 0x2000, "x384d3" },
     /* Xmega, >= 128K FLASH, > 64K RAM.  */
   { "avrxmega7",    ARCH_AVRXMEGA7, NULL,                           0, 0x2000, "x128a1" },
   { "atxmega128a1", ARCH_AVRXMEGA7, "__AVR_ATxmega128A1__",         0, 0x2000, "x128a1" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2011-11-08 15:04:32.000000000 +0530
+++ gcc/config/avr/t-avr	2011-11-08 15:06:12.000000000 +0530
@@ -317,6 +317,7 @@ MULTILIB_MATCHES = \
 	mmcu?avrxmega6=mmcu?atxmega256c3 \
 	mmcu?avrxmega6=mmcu?atxmega256d3 \
 	mmcu?avrxmega6=mmcu?atxmega384c3 \
+	mmcu?avrxmega6=mmcu?atxmega384d3 \
 	mmcu?avrxmega7=mmcu?atxmega128a1 \
 	mmcu?avrxmega7=mmcu?atxmega128a1u \
 	mmcu?avrxmega7=mmcu?atxmega128a4u \
