diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2011-11-16 13:15:05.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2011-11-16 18:19:13.000000000 +0530
@@ -221,6 +221,7 @@ const struct mcu_type_s avr_mcu_types[] 
   { "atmega64rfr2",         ARCH_AVR5, "__AVR_ATmega64RFR2__",      0, 0x0200, "m64rfr2"},
   { "atmega32hvb",          ARCH_AVR5, "__AVR_ATmega32HVB__",       0, 0x0100, "m32hvb" },
   { "atmega32hvbrevb",      ARCH_AVR5, "__AVR_ATmega32HVBREVB__",   1, 0x0100, "m32hvbrevb" },
+  { "atmega48hvf",          ARCH_AVR5, "__AVR_ATmega48HVF__",       0, 0x0100, "m48hvf" }, 
   { "at90can32",            ARCH_AVR5, "__AVR_AT90CAN32__",         0, 0x0100, "can32" },
   { "at90can64",            ARCH_AVR5, "__AVR_AT90CAN64__",         0, 0x0100, "can64" },
   { "at90pwm216",           ARCH_AVR5, "__AVR_AT90PWM216__",        0, 0x0100, "90pwm216" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2011-11-16 13:15:05.000000000 +0530
+++ gcc/config/avr/t-avr	2011-11-16 18:25:27.000000000 +0530
@@ -246,6 +246,7 @@ MULTILIB_MATCHES = \
 	mmcu?avr5=mmcu?atmega16hvbrevb \
 	mmcu?avr5=mmcu?atmega32hvb \
 	mmcu?avr5=mmcu?atmega32hvbrevb \
+	mmcu?avr5=mmcu?atmega48hvf \
 	mmcu?avr5=mmcu?at90can32 \
 	mmcu?avr5=mmcu?at90can64 \
 	mmcu?avr5=mmcu?at90pwm216 \
