diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2011-12-08 14:24:48.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2011-12-08 14:38:53.000000000 +0530
@@ -165,6 +165,7 @@ const struct mcu_type_s avr_mcu_types[] 
   { "atmega16hvbrevb",      ARCH_AVR5, "__AVR_ATmega16HVBREVB__",   1, 0x0100, "m16hvbrevb" },
   { "atmega16m1",           ARCH_AVR5, "__AVR_ATmega16M1__",        0, 0x0100, "m16m1" },
   { "atmega16u4",           ARCH_AVR5, "__AVR_ATmega16U4__",        0, 0x0100, "m16u4" },
+  { "atmega26hvg",          ARCH_AVR5, "__AVR_ATmega26HVG__",       0, 0x0100, "m26hvg" },
   { "atmega32",             ARCH_AVR5, "__AVR_ATmega32__",          0, 0x0060, "m32" },
   { "atmega32a",            ARCH_AVR5, "__AVR_ATmega32A__",         0, 0x0060, "m32a" },
   { "atmega323",            ARCH_AVR5, "__AVR_ATmega323__",         0, 0x0060, "m323" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2011-12-08 14:24:48.000000000 +0530
+++ gcc/config/avr/t-avr	2011-12-08 14:41:10.000000000 +0530
@@ -191,6 +191,7 @@ MULTILIB_MATCHES = \
 	mmcu?avr5=mmcu?atmega169a \
 	mmcu?avr5=mmcu?atmega169p \
 	mmcu?avr5=mmcu?atmega169pa \
+	mmcu?avr5=mmcu?atmega26hvg \
 	mmcu?avr5=mmcu?atmega32 \
 	mmcu?avr5=mmcu?atmega32a \
 	mmcu?avr5=mmcu?atmega323 \
