diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-01-18 11:00:21.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-01-18 11:35:04.000000000 +0530
@@ -267,6 +267,8 @@ const struct mcu_type_s avr_mcu_types[] 
     /* Xmega, <= 8K FLASH.  */
     /* Xmega, > 8K, < 64K FLASH, <= 64K RAM.  */
   { "avrxmega2",    ARCH_AVRXMEGA2, NULL,                           0, 0x2000, "x32a4" },
+  { "atmxt224",     ARCH_AVRXMEGA2, "__AVR_ATMXT224__",             0, 0x2000, "mxt224"},
+  { "atmxt224e",    ARCH_AVRXMEGA2, "__AVR_ATMXT224E__",            0, 0x2000, "mxt224e"},
   { "atmxt336s",    ARCH_AVRXMEGA2, "__AVR_ATMXT336S__",            0, 0x2000, "mxt336s"},
   { "atxmega16a4",  ARCH_AVRXMEGA2, "__AVR_ATxmega16A4__",          0, 0x2000, "x16a4" },
   { "atxmega16a4u",  ARCH_AVRXMEGA2, "__AVR_ATxmega16A4U__",        0, 0x2000, "x16a4u" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-01-18 11:00:21.000000000 +0530
+++ gcc/config/avr/t-avr	2012-01-18 11:36:27.000000000 +0530
@@ -282,6 +282,8 @@ MULTILIB_MATCHES = \
 	mmcu?avr6=mmcu?atmega256rfr2 \
  	mmcu?avr6=mmcu?atmega2560 \
 	mmcu?avr6=mmcu?atmega2561 \
+	mmcu?avrxmega2=mmcu?atmxt224 \
+	mmcu?avrxmega2=mmcu?atmxt224e \
 	mmcu?avrxmega2=mmcu?atmxt336s \
 	mmcu?avrxmega2=mmcu?atxmega16a4 \
 	mmcu?avrxmega2=mmcu?atxmega16a4u \
