diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-02-15 14:00:53.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-02-17 11:04:27.000000000 +0530
@@ -116,6 +116,8 @@ const struct mcu_type_s avr_mcu_types[] 
   { "attiny1634",           ARCH_AVR35, "__AVR_ATtiny1634__",       0, 0x0100, "tn1634" },
     /* Enhanced, <= 8K + 2-byte PC + { MOVW/LPMX, MUL }.  */
   { "avr4",                 ARCH_AVR4, NULL,                        0, 0x0060, "m8" },
+  { "ata6285",              ARCH_AVR4, "__AVR_ATA6285__",           0, 0x0100, "a6285" },
+  { "ata6286",              ARCH_AVR4, "__AVR_ATA6286__",           0, 0x0100, "a6286" },
   { "atmega8",              ARCH_AVR4, "__AVR_ATmega8__",           0, 0x0060, "m8" },
   { "atmega8a",             ARCH_AVR4, "__AVR_ATmega8A__",          0, 0x0060, "m8a" },
   { "atmega48",             ARCH_AVR4, "__AVR_ATmega48__",          0, 0x0100, "m48" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-02-15 14:00:53.000000000 +0530
+++ gcc/config/avr/t-avr	2012-02-17 11:30:56.000000000 +0530
@@ -151,6 +151,8 @@ MULTILIB_MATCHES = \
 	mmcu?avr35=mmcu?attiny167 \
 	mmcu?avr35=mmcu?attiny1634 \
 	mmcu?avr35=mmcu?attiny327 \
+	mmcu?avr4=mmcu?ata6285 \
+	mmcu?avr4=mmcu?ata6286 \
 	mmcu?avr4=mmcu?atmega48 \
 	mmcu?avr4=mmcu?atmega48a \
 	mmcu?avr4=mmcu?atmega48pa \
