diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-02-27 09:26:40.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-02-27 09:27:37.000000000 +0530
@@ -144,6 +144,9 @@ const struct mcu_type_s avr_mcu_types[] 
   { "avr5",                 ARCH_AVR5, NULL,                        0, 0x0060, "m16" },
   { "at90pwm161",           ARCH_AVR5, "__AVR_AT90PWM161__",    1, 0x0100, "90pwm161" },
 
+  { "ata5790",              ARCH_AVR5, "__AVR_ATA5790__",           0, 0x0100, "a5790"}, 
+  { "ata5790n",             ARCH_AVR5, "__AVR_ATA5790N__",          0, 0x0100, "a5790n"}, 
+  { "ata5795",              ARCH_AVR5, "__AVR_ATA5795__",           0, 0x0100, "a5795"}, 
   { "atmega16",             ARCH_AVR5, "__AVR_ATmega16__",          0, 0x0060, "m16" },
   { "atmega16a",            ARCH_AVR5, "__AVR_ATmega16A__",         0, 0x0060, "m16a" },
   { "atmega161",            ARCH_AVR5, "__AVR_ATmega161__",         0, 0x0060, "m161" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-02-27 09:26:40.000000000 +0530
+++ gcc/config/avr/t-avr	2012-02-27 09:27:37.000000000 +0530
@@ -175,6 +175,9 @@ MULTILIB_MATCHES = \
 	mmcu?avr4=mmcu?at90pwm3b \
 	mmcu?avr4=mmcu?at90pwm81 \
 	mmcu?avr5=mmcu?at90pwm161 \
+	mmcu?avr5=mmcu?ata5790 \
+	mmcu?avr5=mmcu?ata5790n \
+	mmcu?avr5=mmcu?ata5795 \
 	mmcu?avr5=mmcu?atmega16 \
 	mmcu?avr5=mmcu?atmega16a \
 	mmcu?avr5=mmcu?atmega161 \
