diff -Naurp gcc/config/avr/avr-devices.c config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-06-28 16:20:39.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-06-28 16:23:31.000000000 +0530
@@ -312,6 +312,8 @@ const struct mcu_type_s avr_mcu_types[] 
   { "atxmega128c3", ARCH_AVRXMEGA6, "__AVR_ATxmega128C3__",         0, 0x2000, "x128c3" },
   { "atxmega128d3", ARCH_AVRXMEGA6, "__AVR_ATxmega128D3__",         0, 0x2000, "x128d3" },
   { "atxmega128d4", ARCH_AVRXMEGA6, "__AVR_ATxmega128D4__",         0, 0x2000, "x128d4" },
+  { "atmxt540s",    ARCH_AVRXMEGA6, "__AVR_ATMXT540S__",            0, 0x2000, "mxt540s"},
+  { "atmxt540sreva", ARCH_AVRXMEGA6, "__AVR_ATMXT540SREVA__",     0, 0x2000, "mxt540sreva"},
   { "atxmega192a3", ARCH_AVRXMEGA6, "__AVR_ATxmega192A3__",         0, 0x2000, "x192a3" },
   { "atxmega192a3u",ARCH_AVRXMEGA6, "__AVR_ATxmega192A3U__",        0, 0x2000, "x192a3u"},
   { "atxmega192c3", ARCH_AVRXMEGA6, "__AVR_ATxmega192C3__",         0, 0x2000, "x192c3" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-06-28 16:20:39.000000000 +0530
+++ gcc/config/avr/t-avr	2012-06-28 16:37:18.000000000 +0530
@@ -318,6 +318,8 @@ MULTILIB_MATCHES = \
 	mmcu?avrxmega6=mmcu?atxmega128c3 \
 	mmcu?avrxmega6=mmcu?atxmega128d3 \
 	mmcu?avrxmega6=mmcu?atxmega128d4 \
+	mmcu?avrxmega6=mmcu?atmxt540s \
+	mmcu?avrxmega6=mmcu?atmxt540sreva \
 	mmcu?avrxmega6=mmcu?atxmega192a3 \
 	mmcu?avrxmega6=mmcu?atxmega192a3u \
 	mmcu?avrxmega6=mmcu?atxmega192c3 \
