diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-06-12 11:41:02.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-06-12 11:48:03.000000000 +0530
@@ -99,6 +99,7 @@ const struct mcu_type_s avr_mcu_types[] 
   { "attiny48",             ARCH_AVR25, "__AVR_ATtiny48__",         0, 0x0100, "tn48" },
   { "attiny88",             ARCH_AVR25, "__AVR_ATtiny88__",         0, 0x0100, "tn88" },
   { "attiny828",            ARCH_AVR25, "__AVR_ATtiny828__",        0, 0x0100, "tn828" },
+  { "attiny841",            ARCH_AVR25, "__AVR_ATtiny841__",        0, 0x0100, "tn841" },
   { "at86rf401",            ARCH_AVR25, "__AVR_AT86RF401__",        1, 0x0060, "86401" },
     /* Classic, > 8K, <= 64K +2-byte PC + { JMP/CALL }.  */
   { "avr3",                 ARCH_AVR3, NULL,                        0, 0x0060, "43355" },
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-06-12 11:41:02.000000000 +0530
+++ gcc/config/avr/t-avr	2012-06-12 11:45:26.000000000 +0530
@@ -140,6 +140,7 @@ MULTILIB_MATCHES = \
 	mmcu?avr25=mmcu?attiny48 \
 	mmcu?avr25=mmcu?attiny88 \
 	mmcu?avr25=mmcu?attiny828 \
+	mmcu?avr25=mmcu?attiny841 \
 	mmcu?avr25=mmcu?at86rf401 \
 	mmcu?avr3=mmcu?at43usb355 \
 	mmcu?avr3=mmcu?at76c711 \
