diff -Naurp gcc/config/avr/avr-devices.c gcc/config/avr/avr-devices.c
--- gcc/config/avr/avr-devices.c	2012-10-16 16:56:11.000000000 +0530
+++ gcc/config/avr/avr-devices.c	2012-10-16 16:55:50.000000000 +0530
@@ -291,6 +291,9 @@ const struct mcu_type_s avr_mcu_types[] 
   { "atxmega32a4u",  ARCH_AVRXMEGA2, "__AVR_ATxmega32A4U__",        0, 0x2000, "x32a4u" },
   { "atxmega32c4",  ARCH_AVRXMEGA2, "__AVR_ATxmega32C4__",          0, 0x2000, "x32c4" },
   { "atxmega32d4",  ARCH_AVRXMEGA2, "__AVR_ATxmega32D4__",          0, 0x2000, "x32d4" },
+  { "atxmega16e5",  ARCH_AVRXMEGA2, "__AVR_ATxmega16E5__",          0, 0x2000, "x16e5" },
+  { "atxmega8e5",   ARCH_AVRXMEGA2, "__AVR_ATxmega8E5__",	     0, 0x2000, "x8e5"  },
+  { "atxmega32e5",  ARCH_AVRXMEGA2, "__AVR_ATxmega32E5__",          0, 0x2000, "x32e5" },
   { "atxmega32x1",  ARCH_AVRXMEGA2, "__AVR_ATxmega32X1__",          0, 0x2000, "x32x1" },
     /* Xmega, > 8K, <= 64K FLASH, > 64K RAM.  */
     /* { "avrxmega3",    ARCH_AVRXMEGA3, NULL }, */
diff -Naurp gcc/config/avr/t-avr gcc/config/avr/t-avr
--- gcc/config/avr/t-avr	2012-10-16 16:56:11.000000000 +0530
+++ gcc/config/avr/t-avr	2012-10-16 16:55:50.000000000 +0530
@@ -300,6 +300,9 @@ MULTILIB_MATCHES = \
 	mmcu?avrxmega2=mmcu?atxmega16d4 \
 	mmcu?avrxmega2=mmcu?atxmega32c4 \
 	mmcu?avrxmega2=mmcu?atxmega32d4 \
+	mmcu?avrxmega2=mmcu?atxmega32e5 \
+	mmcu?avrxmega2=mmcu?atxmega16e5 \
+	mmcu?avrxmega2=mmcu?atxmega8e5 \
 	mmcu?avrxmega2=mmcu?atxmega32a4 \
 	mmcu?avrxmega2=mmcu?atxmega32a4u \
 	mmcu?avrxmega2=mmcu?atxmega32x1 \
