diff -Naurp gcc/config/avr/avr.h gcc/config/avr/avr.h
--- gcc/config/avr/avr.h
+++ gcc/config/avr/avr.h
@@ -627,7 +627,8 @@ extern const char *avr_device_to_sp8 (in
 #define ASM_SPEC "%{mmcu=avr25:-mmcu=avr2;mmcu=avr35:-mmcu=avr3;mmcu=avr31:-mmcu=avr3;mmcu=avr51:-mmcu=avr5;\
 mmcu=*:-mmcu=%*} \
 %{mmcu=*:%{!mmcu=avr2:%{!mmcu=at90s8515:%{!mmcu=avr31:%{!mmcu=atmega103:\
--mno-skip-bug}}}}}"
+-mno-skip-bug}}}}} \
+%{mrelax:-mlink-relax}"

 #define LINK_SPEC "\
 %{mrelax:--relax\
