diff -Naurp gcc/config/avr/avr.c gcc/config/avr/avr.c
--- gcc/config/avr/avr.c	2012-06-27 15:41:23.000000000 +0530
+++ gcc/config/avr/avr.c	2012-06-27 15:42:02.000000000 +0530
@@ -751,15 +751,15 @@ expand_prologue (void)
 	emit_insn (gen_enable_interrupt ());
 	
       /* Push zero reg.  */
-      emit_push_byte (ZERO_REGNO, true);
+      emit_push_byte ((AVR_TINY)?ZERO_REGNO_AVRTINY10:ZERO_REGNO, true);
 
       /* Push tmp reg.  */
-      emit_push_byte (TMP_REGNO, true);
+      emit_push_byte ((AVR_TINY)?TMP_REGNO_AVRTINY10:TMP_REGNO, true);
 
       /* Push SREG.  */
       /* ??? There's no dwarf2 column reserved for SREG.  */
       emit_move_insn (tmp_reg_rtx, gen_rtx_MEM (QImode, GEN_INT (AVR_SREG_ADDR)));
-      emit_push_byte (TMP_REGNO, false);
+      emit_push_byte ((AVR_TINY)?TMP_REGNO_AVRTINY10:TMP_REGNO, false);
 
       /* Push RAMPD, RAMPX, RAMPY. */
       /*
@@ -1217,16 +1217,16 @@ expand_epilogue (void)
 		} 
 
           /* Restore SREG using tmp reg as scratch.  */
-          emit_pop_byte (TMP_REGNO);
+          emit_pop_byte ((AVR_TINY)?TMP_REGNO_AVRTINY10:TMP_REGNO);
       
           emit_move_insn (gen_rtx_MEM (QImode, GEN_INT (AVR_SREG_ADDR)), 
 			  tmp_reg_rtx);
 
           /* Restore tmp REG.  */
-          emit_pop_byte (TMP_REGNO);
+          emit_pop_byte ((AVR_TINY)?TMP_REGNO_AVRTINY10:TMP_REGNO);
 
           /* Restore zero REG.  */
-          emit_pop_byte (ZERO_REGNO);
+          emit_pop_byte ((AVR_TINY)?ZERO_REGNO_AVRTINY10:ZERO_REGNO);
         }
 
       emit_jump_insn (gen_return ());
