--- gcc-4.6.2/libcpp/expr.c~	2011-04-29 00:39:59.000000000 +0200
+++ gcc-4.6.2/libcpp/expr.c	2012-12-07 22:38:30.102991906 +0100
@@ -434,9 +434,9 @@
 		                : N_("use of C99 long long integer constant");
 
 	  if (CPP_OPTION (pfile, c99))
-            cpp_warning (pfile, CPP_W_LONG_LONG, message);
+            cpp_warning (pfile, CPP_W_LONG_LONG, "%s", message);
           else
-            cpp_pedwarning (pfile, CPP_W_LONG_LONG, message);
+            cpp_pedwarning (pfile, CPP_W_LONG_LONG, "%s", message);
         }
 
       result |= CPP_N_INTEGER;
--- gcc-4.6.2/libcpp/macro.c~	2010-09-29 16:49:14.000000000 +0200
+++ gcc-4.6.2/libcpp/macro.c	2012-12-07 22:45:54.119642910 +0100
@@ -1761,7 +1761,7 @@
 	     function-like macros, but not at the end.  */
 	  if (following_paste_op)
 	    {
-	      cpp_error (pfile, CPP_DL_ERROR, paste_op_error_msg);
+	      cpp_error (pfile, CPP_DL_ERROR, "%s", paste_op_error_msg);
 	      return false;
 	    }
 	  break;
@@ -1774,7 +1774,7 @@
 	     function-like macros, but not at the beginning.  */
 	  if (macro->count == 1)
 	    {
-	      cpp_error (pfile, CPP_DL_ERROR, paste_op_error_msg);
+	      cpp_error (pfile, CPP_DL_ERROR, "%s", paste_op_error_msg);
 	      return false;
 	    }
 
--- gcc-4.6.2/gcc/lto-wrapper.c~	2011-02-01 15:42:08.000000000 +0100
+++ gcc-4.6.2/gcc/lto-wrapper.c	2012-12-07 22:52:38.609628642 +0100
@@ -181,10 +181,10 @@
       if (err != 0)
 	{
 	  errno = err;
-	  fatal_perror (errmsg);
+	  fatal_perror ("%s", errmsg);
 	}
       else
-	fatal (errmsg);
+	fatal ("%s", errmsg);
     }
 
   return pex;
--- gcc-4.6.2/gcc/dwarf2asm.c~	2010-12-04 23:01:32.000000000 +0100
+++ gcc-4.6.2/gcc/dwarf2asm.c	2012-12-07 23:00:36.922945101 +0100
@@ -167,7 +167,7 @@
 #ifndef ASM_OUTPUT_DWARF_VMS_DELTA
   /* VMS Delta is only special on ia64-vms, but this funtion also gets
      called on alpha-vms so it has to do something sane.  */
-  dw2_asm_output_delta (size, lab1, lab2, comment);
+  dw2_asm_output_delta (size, lab1, lab2, "%s", comment);
 #else
   ASM_OUTPUT_DWARF_VMS_DELTA (asm_out_file, size, lab1, lab2);
   if (flag_debug_asm && comment)
